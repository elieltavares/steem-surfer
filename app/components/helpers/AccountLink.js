import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { getAccount } from '../../backend/steem-client';

export const makePath = username => `/@${username}`;

class AccountLink extends Component {
  goProfile = async () => {
    const { username, history, actions, onClick, afterClick } = this.props;

    onClick();

    let { accountData } = this.props;
    const { setVisitingAccount } = actions;

    if (!accountData) {
      try {
        accountData = await getAccount(username);
      } catch (err) {
        accountData = null;
      }
    }

    if (accountData) {
      setVisitingAccount(accountData);
    }

    history.push(makePath(username));

    afterClick();
  };

  render() {
    const { children } = this.props;

    return React.cloneElement(children, {
      onClick: this.goProfile
    });
  }
}

AccountLink.defaultProps = {
  accountData: null,
  onClick: () => {},
  afterClick: () => {}
};

AccountLink.propTypes = {
  children: PropTypes.element.isRequired,
  username: PropTypes.string.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  accountData: PropTypes.instanceOf(Object),
  onClick: PropTypes.func,
  afterClick: PropTypes.func,
  actions: PropTypes.shape({
    setVisitingAccount: PropTypes.func.isRequired
  }).isRequired
};

export default AccountLink;
