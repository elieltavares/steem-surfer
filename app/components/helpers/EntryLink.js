import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { getContent } from '../../backend/steem-client';

export const makePath = (category, author, permlink, toReplies = false) =>
  `/${category}/@${author}/${permlink}${toReplies ? '#replies' : ''}`;

class EntryLink extends Component {
  goEntry = async () => {
    let { entry } = this.props;
    const { history, actions, toReplies, markAsRead } = this.props;

    if (markAsRead) {
      markAsRead();
    }

    if (!entry) {
      const { author, permlink } = this.props;
      entry = await getContent(author, permlink);
    }

    if (entry) {
      const { category, author, permlink } = entry;
      actions.setVisitingEntry(entry);
      history.push(makePath(category, author, permlink, toReplies));
    }
  };

  render() {
    const { children } = this.props;

    return React.cloneElement(children, {
      onClick: this.goEntry
    });
  }
}

EntryLink.defaultProps = {
  entry: null,
  toReplies: false,
  markAsRead: null
};

EntryLink.propTypes = {
  children: PropTypes.element.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  entry: PropTypes.shape({
    category: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    permlink: PropTypes.string.isRequired
  }),
  author: PropTypes.string.isRequired,
  permlink: PropTypes.string.isRequired,
  toReplies: PropTypes.bool,
  actions: PropTypes.shape({
    setVisitingEntry: PropTypes.func.isRequired
  }).isRequired,
  markAsRead: PropTypes.func
};

export default EntryLink;
