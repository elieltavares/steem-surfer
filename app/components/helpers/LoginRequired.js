import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';

import { Modal } from 'antd';
import { FormattedMessage, FormattedHTMLMessage } from 'react-intl';

import Login from '../dialogs/Login';

export const checkLogin = (activeAccount, requiredKeys) => {
  // Steem connect login.
  if (activeAccount && activeAccount.type === 'sc') {
    return true;
  }

  // Traditional login. Check all required keys exists in user keys.
  if (
    activeAccount &&
    activeAccount.type === 's' &&
    requiredKeys.every(e => activeAccount.keys[e])
  ) {
    return true;
  }

  return false;
};

class LoginRequired extends Component {
  constructor(props) {
    super(props);

    this.state = {
      loginModalVisible: false
    };
  }

  showLoginModal = () => {
    this.setState({
      loginModalVisible: true
    });

    setTimeout(() => {
      const { onDialogOpen } = this.props;
      onDialogOpen();
    }, 200);
  };

  onLoginModalCancel = () => {
    this.setState({
      loginModalVisible: false
    });
  };

  onLoginSuccess = () => {
    this.setState({
      loginModalVisible: false
    });

    setTimeout(() => {
      if (!this.check()) {
        this.showLoginModal();
      }
    }, 400);
  };

  afterModalClose = () => {
    const { onDialogClose } = this.props;
    onDialogClose();
  };

  check = () => {
    const { activeAccount, requiredKeys } = this.props;

    return checkLogin(activeAccount, requiredKeys);
  };

  render() {
    const { children } = this.props;

    if (this.check()) {
      return children;
    }

    const { activeAccount, requiredKeys } = this.props;
    const { loginModalVisible } = this.state;

    // Default message
    let loginMsg = <FormattedMessage id="login-required.default" />;

    if (activeAccount) {
      // More specific, key based message
      loginMsg =
        requiredKeys.length > 1 ? (
          <FormattedHTMLMessage
            id="login-required.keys-required"
            values={{ keys: requiredKeys }}
          />
        ) : (
          <FormattedHTMLMessage
            id="login-required.key-required"
            values={{ key: requiredKeys }}
          />
        );
    }

    // Clone child element rewriting onclick event
    const newChild = React.cloneElement(children, {
      onClick: () => {
        this.showLoginModal();
      }
    });

    return (
      <Fragment>
        {newChild}

        <Modal
          visible={loginModalVisible}
          onCancel={this.onLoginModalCancel}
          footer={false}
          width="500px"
          closable={false}
          destroyOnClose
          centered
          afterClose={this.afterModalClose}
        >
          <Login
            {...this.props}
            onSuccess={this.onLoginSuccess}
            loginMsg={loginMsg}
          />
        </Modal>
      </Fragment>
    );
  }
}

LoginRequired.defaultProps = {
  activeAccount: null,
  onDialogOpen: () => {},
  onDialogClose: () => {},
  requiredKeys: []
};

LoginRequired.propTypes = {
  children: PropTypes.element.isRequired,
  activeAccount: PropTypes.instanceOf(Object),
  requiredKeys: PropTypes.arrayOf(PropTypes.string),
  onDialogOpen: PropTypes.func,
  onDialogClose: PropTypes.func
};

export default LoginRequired;
