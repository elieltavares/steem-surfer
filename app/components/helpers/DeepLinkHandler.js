import React from 'react';
import PropTypes from 'prop-types';

import filters from '../../constants/filters.json';

import { makePath as makePathTag } from './TagLink';
import { makePath as makePathAccount } from './AccountLink';
import { makePath as makePathEntry } from './EntryLink';

export const deepLink2Obj = url => {
  let urlPart = url.split('://')[1];

  // remove last char if /
  if (urlPart.endsWith('/')) {
    urlPart = urlPart.substring(0, urlPart.length - 1);
  }

  const parts = urlPart.split('/');

  // witnesses
  if (parts.length === 1 && parts[0] === 'witnesses') {
    return { type: 'witnesses' };
  }

  // sps
  if (parts.length === 1 && parts[0] === 'sps') {
    return { type: 'sps' };
  }

  // filter
  if (parts.length === 1 && filters.includes(parts[0])) {
    return { type: 'filter', filter: parts[0] };
  }

  // filter with tag
  if (parts.length === 2 && filters.includes(parts[0])) {
    return { type: 'filter-tag', filter: parts[0], tag: parts[1] };
  }

  // account
  if (parts.length === 1 && parts[0].startsWith('@')) {
    return { type: 'account', account: parts[0].replace('@', '') };
  }

  // post without tag
  if (parts.length === 2 && parts[0].startsWith('@')) {
    return {
      type: 'post',
      cat: 'esteem',
      author: parts[0].replace('@', ''),
      permlink: parts[1]
    };
  }

  // post with tag
  if (parts.length === 3 && parts[1].startsWith('@')) {
    return {
      type: 'post',
      cat: parts[0],
      author: parts[1].replace('@', ''),
      permlink: parts[2]
    };
  }
};

class DeepLinkHandler extends React.Component {
  componentDidMount() {
    window.addEventListener('deep-link', this.handler);
  }

  componentWillUnmount() {
    window.removeEventListener('deep-link', this.handler);
  }

  handler = e => {
    const { history } = this.props;
    const { url } = e.detail;
    const obj = deepLink2Obj(url);

    let path;

    switch (obj.type) {
      case 'filter':
        path = `/${obj.filter}`;
        break;
      case 'filter-tag':
        path = makePathTag(obj.filter, obj.tag);
        break;
      case 'account':
        path = makePathAccount(obj.account);
        break;
      case 'post':
        path = makePathEntry(obj.cat, obj.author, obj.permlink);
        break;
      case 'sps':
        path = `/sps`;
        break;
      case 'witnesses':
        path = `/witnesses`;
        break;
      default:
        path = null;
        break;
    }

    if (path) {
      history.push(path);
    }
  };

  render() {
    return null;
  }
}

DeepLinkHandler.propTypes = {
  history: PropTypes.shape({}).isRequired
};

export default DeepLinkHandler;
