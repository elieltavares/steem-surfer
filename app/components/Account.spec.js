/* eslint-disable */

import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {
  AccountMenu,
  AccountCover,
  AccountTopPosts,
  Exchange
} from './Account';
import { mountWithIntl, shallowWithIntl } from '../helpers/intl-test';

Enzyme.configure({ adapter: new Adapter() });

describe('AccountMenu', () => {
  it('Render', () => {
    const props = {
      username: 'foo',
      section: 'comments',
      history: {},
      global: {
        listStyle: 'row'
      },
      actions: {
        changeListStyle: () => {}
      }
    };
    const wrapper = mountWithIntl(<AccountMenu {...props} />);

    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe('AccountCover', () => {
  it('Render', () => {
    const props = {
      username: 'foo',
      account: {
        accountProfile: {
          cover_image: 'https://foo.bar/baz.png'
        }
      },
      global: {
        theme: 'day'
      }
    };
    const wrapper = mountWithIntl(<AccountCover {...props} />);

    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe('AccountTopPosts', () => {
  it('Render', () => {
    const props = {
      posts: [
        {
          title: 'Foo bar',
          body: 'lorem ipsum',
          author: 'fooox',
          permlink: 'lorem-ipsum'
        },
        {
          title: 'Baz gaz',
          body: 'dolor sit amet',
          author: 'baazx',
          permlink: 'dolor-sit'
        }
      ],
      history: {},
      actions: {
        setVisitingEntry: () => {}
      }
    };
    const wrapper = mountWithIntl(<AccountTopPosts {...props} />);

    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe('Exchange', () => {
  it('Render raising values', () => {
    const props = {
      marketData: {
        sbd: {
          quotes: {
            btc: {
              last_updated: '2018-10-15T08:39:22.000Z',
              percent_change: 0.8654,
              price: 0.00015030939917398168
            },
            usd: {
              last_updated: '2018-10-15T08:39:29.000Z',
              percent_change: 0.719866,
              price: 0.99741902187
            }
          }
        },
        steem: {
          quotes: {
            btc: {
              last_updated: '2018-10-15T08:39:22.000Z',
              percent_change: 1.0734,
              price: 0.0001246760657740764
            },
            usd: {
              last_updated: '2018-10-15T08:39:29.000Z',
              percent_change: 0.511095,
              price: 0.82732204545
            }
          }
        }
      }
    };
    const wrapper = mountWithIntl(<Exchange {...props} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Render falling values', () => {
    const props = {
      marketData: {
        sbd: {
          quotes: {
            btc: {
              last_updated: '2018-10-15T08:39:22.000Z',
              percent_change: -0.8654,
              price: 0.00015030939917398168
            },
            usd: {
              last_updated: '2018-10-15T08:39:29.000Z',
              percent_change: -0.719866,
              price: 0.99741902187
            }
          }
        },
        steem: {
          quotes: {
            btc: {
              last_updated: '2018-10-15T08:39:22.000Z',
              percent_change: -1.0734,
              price: 0.0001246760657740764
            },
            usd: {
              last_updated: '2018-10-15T08:39:29.000Z',
              percent_change: -0.511095,
              price: 0.82732204545
            }
          }
        }
      }
    };
    const wrapper = mountWithIntl(<Exchange {...props} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Render same values', () => {
    const props = {
      marketData: {
        sbd: {
          quotes: {
            btc: {
              last_updated: '2018-10-15T08:39:22.000Z',
              percent_change: 0.0,
              price: 0.00015030939917398168
            },
            usd: {
              last_updated: '2018-10-15T08:39:29.000Z',
              percent_change: 0.0,
              price: 0.99741902187
            }
          }
        },
        steem: {
          quotes: {
            btc: {
              last_updated: '2018-10-15T08:39:22.000Z',
              percent_change: 0.0,
              price: 0.0001246760657740764
            },
            usd: {
              last_updated: '2018-10-15T08:39:29.000Z',
              percent_change: 0.0,
              price: 0.82732204545
            }
          }
        }
      }
    };
    const wrapper = mountWithIntl(<Exchange {...props} />);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
