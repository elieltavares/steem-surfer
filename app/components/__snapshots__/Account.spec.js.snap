// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountCover Render 1`] = `"<div class=\\"account-cover\\"><div class=\\"cover-image\\" style=\\"background-image: url(https://steemitimages.com/0x0/https://foo.bar/baz.png);\\"></div><div class=\\"follow-controls-holder\\"><div><button style=\\"margin-right: 5px;\\" type=\\"button\\" class=\\"ant-btn ant-btn-primary\\"><span>Follow</span></button><button type=\\"button\\" class=\\"ant-btn\\"><span>Mute</span></button></div></div></div>"`;

exports[`AccountMenu Render 1`] = `"<div class=\\"account-menu\\"><div class=\\"account-menu-items\\"><a role=\\"none\\" class=\\"menu-item false\\"><span>Blog</span></a><a role=\\"none\\" class=\\"menu-item selected-item\\"><span>Comments</span></a><a role=\\"none\\" class=\\"menu-item false\\"><span>Replies</span></a><a role=\\"none\\" class=\\"menu-item false\\"><span>Points</span></a><a role=\\"none\\" class=\\"menu-item false\\"><span>Wallet</span></a></div><div class=\\"page-tools\\"><a class=\\"list-switch \\" role=\\"none\\"><i class=\\"mi\\">view_module</i></a></div></div>"`;

exports[`AccountTopPosts Render 1`] = `"<div class=\\"top-posts-list\\"><h2 class=\\"top-posts-list-header\\"><span>Top Posts</span></h2><div class=\\"top-posts-list-body\\"><div class=\\"top-posts-list-item\\" role=\\"none\\"><div class=\\"post-image\\"><img alt=\\"\\" src=\\"test-file-stub\\"></div><div class=\\"post-content\\"><div class=\\"post-title\\">Foo bar</div><div class=\\"post-body\\">lorem ipsum</div></div></div><div class=\\"top-posts-list-item\\" role=\\"none\\"><div class=\\"post-image\\"><img alt=\\"\\" src=\\"test-file-stub\\"></div><div class=\\"post-content\\"><div class=\\"post-title\\">Baz gaz</div><div class=\\"post-body\\">dolor sit amet</div></div></div></div></div>"`;

exports[`Exchange Render falling values 1`] = `"<div class=\\"exchange-data\\"><div class=\\"exchange-title\\"><span class=\\"title-icon \\"><i class=\\"mi\\">show_chart</i></span><h2><span>Exchange</span></h2></div><div class=\\"market-data\\"><div class=\\"market-data-title\\">Steem</div><div class=\\"price change-down\\"><div class=\\"price-usd\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"fa-dollar-sign\\" viewBox=\\"0 0 288 512\\"><path fill=\\"currentColor\\" d=\\"M209.2 233.4l-108-31.6C88.7 198.2 80 186.5 80 173.5c0-16.3 13.2-29.5 29.5-29.5h66.3c12.2 0 24.2 3.7 34.2 10.5 6.1 4.1 14.3 3.1 19.5-2l34.8-34c7.1-6.9 6.1-18.4-1.8-24.5C238 74.8 207.4 64.1 176 64V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48h-2.5C45.8 64-5.4 118.7.5 183.6c4.2 46.1 39.4 83.6 83.8 96.6l102.5 30c12.5 3.7 21.2 15.3 21.2 28.3 0 16.3-13.2 29.5-29.5 29.5h-66.3C100 368 88 364.3 78 357.5c-6.1-4.1-14.3-3.1-19.5 2l-34.8 34c-7.1 6.9-6.1 18.4 1.8 24.5 24.5 19.2 55.1 29.9 86.5 30v48c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-48.2c46.6-.9 90.3-28.6 105.7-72.7 21.5-61.6-14.6-124.8-72.5-141.7z\\"></path></svg> <span>0.83</span></div><div class=\\"price-change\\">(<span>-0.51</span>%)</div><i class=\\"mi change-icon\\">arrow_drop_down</i></div><div class=\\"price change-down\\"><div class=\\"price-btc\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"fa-btc\\" viewBox=\\"0 0 384 512\\"><path fill=\\"currentColor\\" d=\\"M310.204 242.638c27.73-14.18 45.377-39.39 41.28-81.3-5.358-57.351-52.458-76.573-114.85-81.929V0h-48.528v77.203c-12.605 0-25.525.315-38.444.63V0h-48.528v79.409c-17.842.539-38.622.276-97.37 0v51.678c38.314-.678 58.417-3.14 63.023 21.427v217.429c-2.925 19.492-18.524 16.685-53.255 16.071L3.765 443.68c88.481 0 97.37.315 97.37.315V512h48.528v-67.06c13.234.315 26.154.315 38.444.315V512h48.528v-68.005c81.299-4.412 135.647-24.894 142.895-101.467 5.671-61.446-23.32-88.862-69.326-99.89zM150.608 134.553c27.415 0 113.126-8.507 113.126 48.528 0 54.515-85.71 48.212-113.126 48.212v-96.74zm0 251.776V279.821c32.772 0 133.127-9.138 133.127 53.255-.001 60.186-100.355 53.253-133.127 53.253z\\"></path></svg> <span>0.00012468</span></div><div class=\\"price-change\\">(<span>-1.07</span>%)</div><i class=\\"mi change-icon\\">arrow_drop_down</i></div></div><div class=\\"market-data\\"><div class=\\"market-data-title\\">Steem Dollar</div><div class=\\"price change-down\\"><div class=\\"price-usd\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"fa-dollar-sign\\" viewBox=\\"0 0 288 512\\"><path fill=\\"currentColor\\" d=\\"M209.2 233.4l-108-31.6C88.7 198.2 80 186.5 80 173.5c0-16.3 13.2-29.5 29.5-29.5h66.3c12.2 0 24.2 3.7 34.2 10.5 6.1 4.1 14.3 3.1 19.5-2l34.8-34c7.1-6.9 6.1-18.4-1.8-24.5C238 74.8 207.4 64.1 176 64V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48h-2.5C45.8 64-5.4 118.7.5 183.6c4.2 46.1 39.4 83.6 83.8 96.6l102.5 30c12.5 3.7 21.2 15.3 21.2 28.3 0 16.3-13.2 29.5-29.5 29.5h-66.3C100 368 88 364.3 78 357.5c-6.1-4.1-14.3-3.1-19.5 2l-34.8 34c-7.1 6.9-6.1 18.4 1.8 24.5 24.5 19.2 55.1 29.9 86.5 30v48c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-48.2c46.6-.9 90.3-28.6 105.7-72.7 21.5-61.6-14.6-124.8-72.5-141.7z\\"></path></svg> <span>1</span></div><div class=\\"price-change\\">(<span>-0.72</span>%)</div><i class=\\"mi change-icon\\">arrow_drop_down</i></div><div class=\\"price change-down\\"><div class=\\"price-btc\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"fa-btc\\" viewBox=\\"0 0 384 512\\"><path fill=\\"currentColor\\" d=\\"M310.204 242.638c27.73-14.18 45.377-39.39 41.28-81.3-5.358-57.351-52.458-76.573-114.85-81.929V0h-48.528v77.203c-12.605 0-25.525.315-38.444.63V0h-48.528v79.409c-17.842.539-38.622.276-97.37 0v51.678c38.314-.678 58.417-3.14 63.023 21.427v217.429c-2.925 19.492-18.524 16.685-53.255 16.071L3.765 443.68c88.481 0 97.37.315 97.37.315V512h48.528v-67.06c13.234.315 26.154.315 38.444.315V512h48.528v-68.005c81.299-4.412 135.647-24.894 142.895-101.467 5.671-61.446-23.32-88.862-69.326-99.89zM150.608 134.553c27.415 0 113.126-8.507 113.126 48.528 0 54.515-85.71 48.212-113.126 48.212v-96.74zm0 251.776V279.821c32.772 0 133.127-9.138 133.127 53.255-.001 60.186-100.355 53.253-133.127 53.253z\\"></path></svg> <span>0.00015031</span></div><div class=\\"price-change\\">(<span>-0.87</span>%)</div><i class=\\"mi change-icon\\">arrow_drop_down</i></div></div></div>"`;

exports[`Exchange Render raising values 1`] = `"<div class=\\"exchange-data\\"><div class=\\"exchange-title\\"><span class=\\"title-icon \\"><i class=\\"mi\\">show_chart</i></span><h2><span>Exchange</span></h2></div><div class=\\"market-data\\"><div class=\\"market-data-title\\">Steem</div><div class=\\"price change-up\\"><div class=\\"price-usd\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"fa-dollar-sign\\" viewBox=\\"0 0 288 512\\"><path fill=\\"currentColor\\" d=\\"M209.2 233.4l-108-31.6C88.7 198.2 80 186.5 80 173.5c0-16.3 13.2-29.5 29.5-29.5h66.3c12.2 0 24.2 3.7 34.2 10.5 6.1 4.1 14.3 3.1 19.5-2l34.8-34c7.1-6.9 6.1-18.4-1.8-24.5C238 74.8 207.4 64.1 176 64V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48h-2.5C45.8 64-5.4 118.7.5 183.6c4.2 46.1 39.4 83.6 83.8 96.6l102.5 30c12.5 3.7 21.2 15.3 21.2 28.3 0 16.3-13.2 29.5-29.5 29.5h-66.3C100 368 88 364.3 78 357.5c-6.1-4.1-14.3-3.1-19.5 2l-34.8 34c-7.1 6.9-6.1 18.4 1.8 24.5 24.5 19.2 55.1 29.9 86.5 30v48c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-48.2c46.6-.9 90.3-28.6 105.7-72.7 21.5-61.6-14.6-124.8-72.5-141.7z\\"></path></svg> <span>0.83</span></div><div class=\\"price-change\\">(<span>0.51</span>%)</div><i class=\\"mi change-icon\\">arrow_drop_up</i></div><div class=\\"price change-up\\"><div class=\\"price-btc\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"fa-btc\\" viewBox=\\"0 0 384 512\\"><path fill=\\"currentColor\\" d=\\"M310.204 242.638c27.73-14.18 45.377-39.39 41.28-81.3-5.358-57.351-52.458-76.573-114.85-81.929V0h-48.528v77.203c-12.605 0-25.525.315-38.444.63V0h-48.528v79.409c-17.842.539-38.622.276-97.37 0v51.678c38.314-.678 58.417-3.14 63.023 21.427v217.429c-2.925 19.492-18.524 16.685-53.255 16.071L3.765 443.68c88.481 0 97.37.315 97.37.315V512h48.528v-67.06c13.234.315 26.154.315 38.444.315V512h48.528v-68.005c81.299-4.412 135.647-24.894 142.895-101.467 5.671-61.446-23.32-88.862-69.326-99.89zM150.608 134.553c27.415 0 113.126-8.507 113.126 48.528 0 54.515-85.71 48.212-113.126 48.212v-96.74zm0 251.776V279.821c32.772 0 133.127-9.138 133.127 53.255-.001 60.186-100.355 53.253-133.127 53.253z\\"></path></svg> <span>0.00012468</span></div><div class=\\"price-change\\">(<span>1.07</span>%)</div><i class=\\"mi change-icon\\">arrow_drop_up</i></div></div><div class=\\"market-data\\"><div class=\\"market-data-title\\">Steem Dollar</div><div class=\\"price change-up\\"><div class=\\"price-usd\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"fa-dollar-sign\\" viewBox=\\"0 0 288 512\\"><path fill=\\"currentColor\\" d=\\"M209.2 233.4l-108-31.6C88.7 198.2 80 186.5 80 173.5c0-16.3 13.2-29.5 29.5-29.5h66.3c12.2 0 24.2 3.7 34.2 10.5 6.1 4.1 14.3 3.1 19.5-2l34.8-34c7.1-6.9 6.1-18.4-1.8-24.5C238 74.8 207.4 64.1 176 64V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48h-2.5C45.8 64-5.4 118.7.5 183.6c4.2 46.1 39.4 83.6 83.8 96.6l102.5 30c12.5 3.7 21.2 15.3 21.2 28.3 0 16.3-13.2 29.5-29.5 29.5h-66.3C100 368 88 364.3 78 357.5c-6.1-4.1-14.3-3.1-19.5 2l-34.8 34c-7.1 6.9-6.1 18.4 1.8 24.5 24.5 19.2 55.1 29.9 86.5 30v48c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-48.2c46.6-.9 90.3-28.6 105.7-72.7 21.5-61.6-14.6-124.8-72.5-141.7z\\"></path></svg> <span>1</span></div><div class=\\"price-change\\">(<span>0.72</span>%)</div><i class=\\"mi change-icon\\">arrow_drop_up</i></div><div class=\\"price change-up\\"><div class=\\"price-btc\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"fa-btc\\" viewBox=\\"0 0 384 512\\"><path fill=\\"currentColor\\" d=\\"M310.204 242.638c27.73-14.18 45.377-39.39 41.28-81.3-5.358-57.351-52.458-76.573-114.85-81.929V0h-48.528v77.203c-12.605 0-25.525.315-38.444.63V0h-48.528v79.409c-17.842.539-38.622.276-97.37 0v51.678c38.314-.678 58.417-3.14 63.023 21.427v217.429c-2.925 19.492-18.524 16.685-53.255 16.071L3.765 443.68c88.481 0 97.37.315 97.37.315V512h48.528v-67.06c13.234.315 26.154.315 38.444.315V512h48.528v-68.005c81.299-4.412 135.647-24.894 142.895-101.467 5.671-61.446-23.32-88.862-69.326-99.89zM150.608 134.553c27.415 0 113.126-8.507 113.126 48.528 0 54.515-85.71 48.212-113.126 48.212v-96.74zm0 251.776V279.821c32.772 0 133.127-9.138 133.127 53.255-.001 60.186-100.355 53.253-133.127 53.253z\\"></path></svg> <span>0.00015031</span></div><div class=\\"price-change\\">(<span>0.87</span>%)</div><i class=\\"mi change-icon\\">arrow_drop_up</i></div></div></div>"`;

exports[`Exchange Render same values 1`] = `"<div class=\\"exchange-data\\"><div class=\\"exchange-title\\"><span class=\\"title-icon \\"><i class=\\"mi\\">show_chart</i></span><h2><span>Exchange</span></h2></div><div class=\\"market-data\\"><div class=\\"market-data-title\\">Steem</div><div class=\\"price change-same\\"><div class=\\"price-usd\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"fa-dollar-sign\\" viewBox=\\"0 0 288 512\\"><path fill=\\"currentColor\\" d=\\"M209.2 233.4l-108-31.6C88.7 198.2 80 186.5 80 173.5c0-16.3 13.2-29.5 29.5-29.5h66.3c12.2 0 24.2 3.7 34.2 10.5 6.1 4.1 14.3 3.1 19.5-2l34.8-34c7.1-6.9 6.1-18.4-1.8-24.5C238 74.8 207.4 64.1 176 64V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48h-2.5C45.8 64-5.4 118.7.5 183.6c4.2 46.1 39.4 83.6 83.8 96.6l102.5 30c12.5 3.7 21.2 15.3 21.2 28.3 0 16.3-13.2 29.5-29.5 29.5h-66.3C100 368 88 364.3 78 357.5c-6.1-4.1-14.3-3.1-19.5 2l-34.8 34c-7.1 6.9-6.1 18.4 1.8 24.5 24.5 19.2 55.1 29.9 86.5 30v48c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-48.2c46.6-.9 90.3-28.6 105.7-72.7 21.5-61.6-14.6-124.8-72.5-141.7z\\"></path></svg> <span>0.83</span></div><div class=\\"price-change\\">(<span>0</span>%)</div><i class=\\"change-icon\\">--</i></div><div class=\\"price change-same\\"><div class=\\"price-btc\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"fa-btc\\" viewBox=\\"0 0 384 512\\"><path fill=\\"currentColor\\" d=\\"M310.204 242.638c27.73-14.18 45.377-39.39 41.28-81.3-5.358-57.351-52.458-76.573-114.85-81.929V0h-48.528v77.203c-12.605 0-25.525.315-38.444.63V0h-48.528v79.409c-17.842.539-38.622.276-97.37 0v51.678c38.314-.678 58.417-3.14 63.023 21.427v217.429c-2.925 19.492-18.524 16.685-53.255 16.071L3.765 443.68c88.481 0 97.37.315 97.37.315V512h48.528v-67.06c13.234.315 26.154.315 38.444.315V512h48.528v-68.005c81.299-4.412 135.647-24.894 142.895-101.467 5.671-61.446-23.32-88.862-69.326-99.89zM150.608 134.553c27.415 0 113.126-8.507 113.126 48.528 0 54.515-85.71 48.212-113.126 48.212v-96.74zm0 251.776V279.821c32.772 0 133.127-9.138 133.127 53.255-.001 60.186-100.355 53.253-133.127 53.253z\\"></path></svg> <span>0.00012468</span></div><div class=\\"price-change\\">(<span>0</span>%)</div><i class=\\"change-icon\\">--</i></div></div><div class=\\"market-data\\"><div class=\\"market-data-title\\">Steem Dollar</div><div class=\\"price change-same\\"><div class=\\"price-usd\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"fa-dollar-sign\\" viewBox=\\"0 0 288 512\\"><path fill=\\"currentColor\\" d=\\"M209.2 233.4l-108-31.6C88.7 198.2 80 186.5 80 173.5c0-16.3 13.2-29.5 29.5-29.5h66.3c12.2 0 24.2 3.7 34.2 10.5 6.1 4.1 14.3 3.1 19.5-2l34.8-34c7.1-6.9 6.1-18.4-1.8-24.5C238 74.8 207.4 64.1 176 64V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48h-2.5C45.8 64-5.4 118.7.5 183.6c4.2 46.1 39.4 83.6 83.8 96.6l102.5 30c12.5 3.7 21.2 15.3 21.2 28.3 0 16.3-13.2 29.5-29.5 29.5h-66.3C100 368 88 364.3 78 357.5c-6.1-4.1-14.3-3.1-19.5 2l-34.8 34c-7.1 6.9-6.1 18.4 1.8 24.5 24.5 19.2 55.1 29.9 86.5 30v48c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-48.2c46.6-.9 90.3-28.6 105.7-72.7 21.5-61.6-14.6-124.8-72.5-141.7z\\"></path></svg> <span>1</span></div><div class=\\"price-change\\">(<span>0</span>%)</div><i class=\\"change-icon\\">--</i></div><div class=\\"price change-same\\"><div class=\\"price-btc\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"fa-btc\\" viewBox=\\"0 0 384 512\\"><path fill=\\"currentColor\\" d=\\"M310.204 242.638c27.73-14.18 45.377-39.39 41.28-81.3-5.358-57.351-52.458-76.573-114.85-81.929V0h-48.528v77.203c-12.605 0-25.525.315-38.444.63V0h-48.528v79.409c-17.842.539-38.622.276-97.37 0v51.678c38.314-.678 58.417-3.14 63.023 21.427v217.429c-2.925 19.492-18.524 16.685-53.255 16.071L3.765 443.68c88.481 0 97.37.315 97.37.315V512h48.528v-67.06c13.234.315 26.154.315 38.444.315V512h48.528v-68.005c81.299-4.412 135.647-24.894 142.895-101.467 5.671-61.446-23.32-88.862-69.326-99.89zM150.608 134.553c27.415 0 113.126-8.507 113.126 48.528 0 54.515-85.71 48.212-113.126 48.212v-96.74zm0 251.776V279.821c32.772 0 133.127-9.138 133.127 53.255-.001 60.186-100.355 53.253-133.127 53.253z\\"></path></svg> <span>0.00015031</span></div><div class=\\"price-change\\">(<span>0</span>%)</div><i class=\\"change-icon\\">--</i></div></div></div>"`;
